Automattic/jetpack
# Donations Block: Update donations block to inner blocks

Fixes Automattic/wp-calypso#17232, Automattic/wp-calypso#17225 and Automattic/wp-calypso#17231. WIP PR (Closed) [22470](https://github.com/Automattic/jetpack/pull/22470)

#### Changes proposed in this Pull Request:

* Ported the current Donations block from using RichText components to use InnerBlocks instead.
* Since the components are now Blocks, we can add style customizations to all the elements of the donations block.
* Created a deprecation and migration for the previous block.
* In order to be able to render the previous block when it hasn't been upgraded to the new version I've needed to inspect the previous block so that we could delegate rendering in the FE to the previous render_callback instead of using the new render_callback (projects/plugins/jetpack/class.jetpack-gutenberg.php)
* This PR is the foundation for our future work on the Donations Block, at the moment we have a template lock so that our user can't leave the block in a state in which donations would not be possible. Also when pressing enter on the Text blocks we will not be creating a new block.

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:






We should test as much as possible, even though the block might look like the previous block, most of the code has been changed. It is a good idea to focus on testing the following functional areas:
- Editor and Styling of all the visual elements work as expected.
- FE view of the Donations Block works as expected.
- Deprecation and Migration of the previous block. We also need to test the rendering and functionality of the previous block on the FE.

##### Testing Editor & Styling
* Edit or create a post.
* Search for the Donations Editable Blocks component and add it to the post.
* Modify all the blocks and styles of the page.
* Check the block navigator and verify that all the icons are correct.
* Check the block navigator and verify that all the blocks are correct.
* Toggle all of the Donations Block controls and check on the FE that they get applied.
* Publish the post and go to the FE, the styles should be applied as in the editor.

##### Testing FE Donations Block
* Open a post that has a Donations Block.
* Navigating through tabs should work.
* The donation button should work and donate monthly, annually and one time, depending on the selected tab.
* The amount selector buttons should work, try to donate that amount.
* The custom amount field should work, try to donate that amount.

##### Testing Deprecation & Migration
* Create a post with the previous version of the Donations Block. (you can do this by un-sandboxing your site).
* Now open the same post but have enabled the new version of the Donations Block. (sandbox your site again).
* The editor should not be broken.
* The configured amounts, tabs, and texts should be the same.
* Do not update the block and navigate to the FE of the post.
* The old Donations Block should be 100% functional.
* Go back to the editor and update the block.
* In the FE the Donations Block should work as expected.

##### Test Evidence
https://user-images.githubusercontent.com/1989914/152847549-87ecd43a-cea5-49a0-81f8-c0c49d155937.mp4
